File: JAVA/GestionRelationEmployees/src/main/java/models/ProjectTask.java
package models;

import java.sql.Date;
import java.sql.Timestamp;

/**
 * Modèle représentant une tâche d'un projet (pour Kanban)
 */
public class ProjectTask {

    // Enums
    public enum Status {
        todo, in_progress, review, done
    }

    public enum Priority {
        low, medium, high
    }

    private int id;
    private int projectId;
    private Integer assignedTo;  // employee_id, peut être null
    private String title;
    private String description;
    private Status status;
    private Priority priority;
    private Integer estimatedHours;
    private int actualHours;
    private Date dueDate;
    private Date completedDate;
    private int orderIndex;  // Pour le tri dans le Kanban
    private Timestamp createdAt;
    private Timestamp updatedAt;

    // Champs transients pour l'affichage
    private String projectName;
    private String assignedToName;

    // ─── Constructeurs ───────────────────────────────────────────────

    public ProjectTask() {}

    public ProjectTask(int projectId, String title, String description,
                       Priority priority, Integer estimatedHours, Date dueDate) {
        this.projectId = projectId;
        this.title = title;
        this.description = description;
        this.priority = priority;
        this.estimatedHours = estimatedHours;
        this.dueDate = dueDate;
        this.status = Status.todo;
        this.actualHours = 0;
        this.orderIndex = 0;
    }

    // ─── Getters & Setters ───────────────────────────────────────────

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public int getProjectId() { return projectId; }
    public void setProjectId(int projectId) { this.projectId = projectId; }

    public Integer getAssignedTo() { return assignedTo; }
    public void setAssignedTo(Integer assignedTo) { this.assignedTo = assignedTo; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public Status getStatus() { return status; }
    public void setStatus(Status status) { this.status = status; }

    public Priority getPriority() { return priority; }
    public void setPriority(Priority priority) { this.priority = priority; }

    public Integer getEstimatedHours() { return estimatedHours; }
    public void setEstimatedHours(Integer estimatedHours) { this.estimatedHours = estimatedHours; }

    public int getActualHours() { return actualHours; }
    public void setActualHours(int actualHours) { this.actualHours = actualHours; }

    public Date getDueDate() { return dueDate; }
    public void setDueDate(Date dueDate) { this.dueDate = dueDate; }

    public Date getCompletedDate() { return completedDate; }
    public void setCompletedDate(Date completedDate) { this.completedDate = completedDate; }

    public int getOrderIndex() { return orderIndex; }
    public void setOrderIndex(int orderIndex) { this.orderIndex = orderIndex; }

    public Timestamp getCreatedAt() { return createdAt; }
    public void setCreatedAt(Timestamp createdAt) { this.createdAt = createdAt; }

    public Timestamp getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(Timestamp updatedAt) { this.updatedAt = updatedAt; }

    // Champs transients
    public String getProjectName() { return projectName; }
    public void setProjectName(String projectName) { this.projectName = projectName; }

    public String getAssignedToName() { return assignedToName; }
    public void setAssignedToName(String assignedToName) { this.assignedToName = assignedToName; }

    // ─── Méthodes utilitaires ────────────────────────────────────────

    /**
     * Retourne un emoji de statut pour l'affichage
     */
    public String getStatusEmoji() {
        return switch (status) {
            case todo -> "□";
            case in_progress -> "⏳";
            case review -> "👁️";
            case done -> "✅";
        };
    }

    /**
     * Retourne un emoji de priorité pour l'affichage
     */
    public String getPriorityEmoji() {
        return switch (priority) {
            case low -> "🟢";
            case medium -> "🟡";
            case high -> "🔴";
        };
    }

    /**
     * Retourne le nom du statut pour l'affichage
     */
    public String getStatusLabel() {
        return switch (status) {
            case todo -> "À faire";
            case in_progress -> "En cours";
            case review -> "Review";
            case done -> "Terminé";
        };
    }

    /**
     * Vérifie si la tâche est en retard
     */
    public boolean isOverdue() {
        if (dueDate == null || status == Status.done) return false;
        return new java.util.Date().after(dueDate);
    }

    /**
     * Retourne le temps restant avant l'échéance
     */
    public String getDaysUntilDue() {
        if (dueDate == null) return "Pas d'échéance";
        if (status == Status.done) return "Terminé";

        long diff = dueDate.getTime() - System.currentTimeMillis();
        long days = diff / (1000 * 60 * 60 * 24);

        if (days < 0) return "⚠️ " + Math.abs(days) + " jours de retard";
        if (days == 0) return "🔥 Aujourd'hui";
        if (days == 1) return "⏰ Demain";
        return days + " jours";
    }

    /**
     * Calcule le taux d'utilisation du temps estimé
     */
    public double getTimeUsageRate() {
        if (estimatedHours == null || estimatedHours == 0) return 0;
        return (double) actualHours / estimatedHours * 100;
    }

    @Override
    public String toString() {
        return "ProjectTask{id=" + id + ", title='" + title + "', status=" + status +
                ", assignedTo=" + assignedToName + "}";
    }
}

File: src/Entity/Projet/ProjectTask.php
<?php

namespace App\Entity\Projet;

use App\Repository\Projet\ProjectTaskRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: ProjectTaskRepository::class)]
#[ORM\Table(name: 'project_tasks')]
#[ORM\Index(name: 'idx_project_tasks_project_id', columns: ['project_id'])]
#[ORM\Index(name: 'idx_project_tasks_assigned_to', columns: ['assigned_to'])]
#[ORM\Index(name: 'idx_project_tasks_status', columns: ['status'])]
#[ORM\Index(name: 'idx_project_tasks_due_date', columns: ['due_date'])]
#[ORM\HasLifecycleCallbacks]
class ProjectTask
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    /** @phpstan-ignore-next-line */
    private ?int $id = null;

    #[ORM\Column(name: 'project_id', type: Types::INTEGER)]
    private ?int $projectId = null;

    #[ORM\Column(name: 'assigned_to', type: Types::INTEGER, nullable: true)]
    private ?int $assignedTo = null;

    #[ORM\Column(length: 255)]
    private ?string $title = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $description = null;

    #[ORM\Column(length: 30)]
    private ?string $status = null;

    #[ORM\Column(length: 20)]
    private ?string $priority = null;

    #[ORM\Column(name: 'estimated_hours', type: Types::INTEGER, nullable: true)]
    private ?int $estimatedHours = null;

    #[ORM\Column(name: 'actual_hours', type: Types::INTEGER)]
    private int $actualHours = 0;

    #[ORM\Column(name: 'due_date', type: Types::DATE_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $dueDate = null;

    #[ORM\Column(name: 'completed_date', type: Types::DATE_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $completedDate = null;

    #[ORM\Column(name: 'order_index', type: Types::INTEGER)]
    private int $orderIndex = 0;

    #[ORM\Column(name: 'created_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $createdAt = null;

    #[ORM\Column(name: 'updated_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $updatedAt = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getProjectId(): ?int
    {
        return $this->projectId;
    }

    public function setProjectId(int $projectId): static
    {
        $this->projectId = $projectId;
        return $this;
    }

    public function getAssignedTo(): ?int
    {
        return $this->assignedTo;
    }

    public function setAssignedTo(?int $assignedTo): static
    {
        $this->assignedTo = $assignedTo;
        return $this;
    }

    public function getTitle(): ?string
    {
        return $this->title;
    }

    public function setTitle(string $title): static
    {
        $this->title = $title;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): static
    {
        $this->description = $description;
        return $this;
    }

    public function getStatus(): ?string
    {
        return $this->status;
    }

    public function setStatus(string $status): static
    {
        $this->status = $status;
        return $this;
    }

    public function getPriority(): ?string
    {
        return $this->priority;
    }

    public function setPriority(string $priority): static
    {
        $this->priority = $priority;
        return $this;
    }

    public function getEstimatedHours(): ?int
    {
        return $this->estimatedHours;
    }

    public function setEstimatedHours(?int $estimatedHours): static
    {
        $this->estimatedHours = $estimatedHours;
        return $this;
    }

    public function getActualHours(): int
    {
        return $this->actualHours;
    }

    public function setActualHours(int $actualHours): static
    {
        $this->actualHours = $actualHours;
        return $this;
    }

    public function getDueDate(): ?\DateTimeInterface
    {
        return $this->dueDate;
    }

    public function setDueDate(?\DateTimeInterface $dueDate): static
    {
        $this->dueDate = $dueDate;
        return $this;
    }

    public function getCompletedDate(): ?\DateTimeInterface
    {
        return $this->completedDate;
    }

    public function setCompletedDate(?\DateTimeInterface $completedDate): static
    {
        $this->completedDate = $completedDate;
        return $this;
    }

    public function getOrderIndex(): int
    {
        return $this->orderIndex;
    }

    public function setOrderIndex(int $orderIndex): static
    {
        $this->orderIndex = $orderIndex;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): ?\DateTimeInterface
    {
        return $this->updatedAt;
    }

    #[ORM\PrePersist]
    public function onPrePersist(): void
    {
        $now = new \DateTime();
        $this->createdAt = $now;
        $this->updatedAt = $now;
    }

    #[ORM\PreUpdate]
    public function onPreUpdate(): void
    {
        $this->updatedAt = new \DateTime();
    }
}

